pkg/linter/linter.go

```diff
@@ -206,7 +206,12 @@ func (l Linter) Lint(ctx context.Context) (report.Report, error) {
 
 	aggregate.Violations = append(aggregate.Violations, goReport.Violations...)
 
-	regoReport, err := l.lintWithRegoRules(ctx, input)
+	aggregates, err := l.collectAggregates(ctx, input)
+	if err != nil {
+		return report.Report{}, fmt.Errorf("failed to collect aggregates using Rego rules: %w", err)
+	}
+
+	regoReport, err := l.lintWithRegoRules(ctx, input, aggregates)
```

Comment:
question: I'm missing something I think. What I see is that the aggregates rules are collected (L209) before being passed to `l.lintWithRegoRules`. This in turn passes them to `evalAndCollect`. These are then all collected onto the report.

What wasn't clear to me in this implementation was when aggregate rules were evaluated on the results of other rules.

In my head, I can't see where step 4 from https://github.com/StyraInc/regal/issues/282#issuecomment-1708528606 is being done.

What I'm not sure about is if I've missed this, or if it's coming in another commit! 😅

Reply:
> What wasn't clear to me in this implementation was when aggregate rules were evaluated on the results of other rules.

They are evaluated only if a violation rule reads the aggregates. This is the file I'm using for my local testing:

```
# METADATA
# description: Could not find a rule satisfying query pattern (some_package).public.export
package custom.regal.rules.acmecorp["require-export-rule"]

import future.keywords
import data.regal.result

# First, let's collect which files satisfy the rule.
aggregate contains entry if {
    input["package"].path[2].value == "public"
    some i
	input["rules"][i].head.name == "export"
	input["rules"][i].head.assign == true
	input["rules"][i].head.value.type == "array"
    entry := { "export_rule_found" : input.regal.file.name }
}

report contains violation if {
    not there_is_at_least_one_export_rule
    violation := result.fail(rego.metadata.chain(), {})
}

there_is_at_least_one_export_rule {
	input.aggregate[_].export_rule_found
}
```

Notes: 
* This could be simplified by simple checking if we have >0 aggregates. I wanted to check the negation as a way to see how one could have multiple aggregates collecting different data, and having different violation rules not interfering with each other.
* The aggregate record is an object, which makes it very flexible. One could abuse it, though. For example:

```
# Danger zone below - just to should the potential
aggregate contains entry if {
    # A way to collect all ASTs, providing the author understands the performance limits and risks.
    entry := { "AST" : input }
}
report contains violation if {
    input.aggregate[_].AST["package"].path[2].value == "public"
    violation := result.fail(rego.metadata.chain(), {})
}
```

Reply:
> They are evaluated only if a violation rule reads the aggregates

Ahh 🤦, gotcha! I knew there was something I was missing.

Getting a test in with that example make sense as a good next step to me.

> One could abuse it, though

I think we can live with that. We're not likely to include rules in upstream Regal of that format, but if people want to write and run their own rules that seems ok.

Reply:
From your example,

```rego
entry := { "export_rule_found" : input.regal.file.name }
```

It'd be good if we had an `result.aggregate` function added which always included `input.regal.file.name` in the returned object, so that we'd have a unique value per aggregate object. If we don't there's a risk they'll be munged as a single value when added to the set. But that's an optimization we can save for later too :)